File: filter_pca.py
#!/usr/bin/env python
import numpy as np
import argparse
import pandas as pd
import sys
# import tensorflow as tf
import random
from sklearn.base import BaseEstimator, RegressorMixin

class FilterMatrix():
    def __init__(self):
        self.matrix = {}
    def lookup(self,c1,c2):
        return self.matrix.get(c1,{}).get(c2)
    def set(self,c1,c2,v):
        self.matrix.setdefault(c1,{})[c2] = v


"""
objective function = sum(y_ - (mu + r_i + c_j + sum(r_ik*c_ik)))**2
err = y_ - (mu + r_i + c_j + sum(r_ik*c_ik))
grad = -1 * err
"""

class FilterPca(BaseEstimator, RegressorMixin):
    def __init__(self, feature_cnt, N1, N2, minval = None, maxval = None):
        self.feature_cnt = feature_cnt
        self.N1 = N1
        self.N2 = N2
        self.is_setup = False
        self.row_features = None
        self.col_features = None
        self.row_offsets = None
        self.col_offsets = None
        self.minval = minval
        self.maxval = maxval
    def fit(self, X, y, iter_cnt):
        """
        args:
            X = [(x,y)]
            X is a list of (row,col) indices. The row, col numbers must be >= 0.

            y = [val]
        """
        self.total_mean = mean(y)
        y = [(v - self.total_mean) for v in y]

        #column means
        col_cnts = [0] * self.N2
        col_sums = [0] * self.N2
        for (_,q),v in zip(X,y):
            col_cnts[q] += 1
            col_sums[q] += v
        self.col_means = [0] * self.N2
        for j in range(self.N2):
            if col_cnts[j] != 0:
                self.col_means[j] = col_sums[j] / float(col_cnts[j])


        #row means
        row_cnts = [0] * self.N1
        row_sums = [0] * self.N1
        for (q,_),v in zip(X,y):
            row_cnts[q] += 1
            row_sums[q] += v
        self.row_means = [0] * self.N1
        for i in range(self.N1):
            if row_cnts[i] != 0:
                self.row_means[i] = row_sums[i] / float(row_cnts[i])


        #start all col_offsets and row_offsets at half of their respective means
        if self.row_offsets is None:
            self.row_offsets = [x/2.0 for x in self.row_means]
        if self.col_offsets is None:
            self.col_offsets = [x/2.0 for x in self.col_means]

        # print(self.total_mean);
        # print(self.col_means);
        if self.row_features is None:
            self.row_features = [[np.random.normal(scale=0.01, size=1)[0] for i in range(self.N1)] for k in range(self.feature_cnt)]
        if self.col_features is None:
            self.col_features = [[np.random.normal(scale=0.01, size=1)[0] for i in range(self.N2)] for k in range(self.feature_cnt)]
        self.lrate = 0.001
        self.lam = 15
        print("starting loop")
        for it in range(iter_cnt):
            print(f"Iteration {it}")
            if it%1 == 0: #MUST: lower frequency, this will slow down a serious run
                sum_sq_err = 0
                for (i,j),v in zip(X,y):
                    sum_sq_err += (v - self.row_offsets[i] - self.col_offsets[j] - sum([self.row_features[k][i]*self.col_features[k][j] for k in range(self.feature_cnt)]))**2
                print("sum_sq_err: " + str(sum_sq_err))
                row_offset_regularization = sum(self.lam * self.row_offsets[i] ** 2 for i in range(self.N1))
                row_feature_regularization = sum(self.lam * (self.row_features[f][i] ** 2) for i in range(self.N1) for f in range(self.feature_cnt))
                col_offset_regularization = sum(self.lam * self.col_offsets[j] ** 2 for j in range(self.N2))
                col_feature_regularization = sum(self.lam * (self.col_features[f][j] ** 2) for j in range(self.N2) for f in range(self.feature_cnt))
                objective_function = sum_sq_err + row_offset_regularization + row_feature_regularization + col_offset_regularization + col_feature_regularization

                print(f"obj_fn: {objective_function}")
            for (i,j),v in zip(X,y):
                err = v - self.row_offsets[i] - self.col_offsets[j] - sum([self.row_features[k][i]*self.col_features[k][j] for k in range(self.feature_cnt)])
                self.row_offsets[i] += self.lrate * err
                self.col_offsets[j] += self.lrate * err
                for f in range(self.feature_cnt):
                    self.row_features[f][i] += self.lrate * err * self.col_features[f][j]
                    self.col_features[f][j] += self.lrate * err * self.row_features[f][i]
            #regularization penalties
            for i in range(self.N1):
                self.row_offsets[i] -= self.lrate * self.lam * self.row_offsets[i]
                for f in range(self.feature_cnt):
                    self.row_features[f][i] -= self.lrate * self.lam * self.row_features[f][i]
            for j in range(self.N2):
                self.col_offsets[j] -= self.lrate * self.lam * self.col_offsets[j]
                for f in range(self.feature_cnt):
                    self.col_features[f][j] -= self.lrate * self.lam * self.col_features[f][j]

    def predict(self, X, use_features=True):
        if use_features:
            raw_score = [self.total_mean + self.row_offsets[i] + self.col_offsets[j] + sum([self.row_features[k][i]*self.col_features[k][j] for k in range(self.feature_cnt)]) for i,j in X]
        else:
            raw_score = [self.total_mean + self.row_offsets[i] + self.col_offsets[j] for i,j in X]
        if self.minval is not None:
            raw_score = [max(self.minval, s) for s in raw_score]
        if self.maxval is not None:
            raw_score = [min(self.maxval, s) for s in raw_score]
        return raw_score
    def score(self, X, y):
        preds = self.predict(X)
        partial_preds = self.predict(X, use_features=False)
        r2 = sum((self.total_mean - val)**2 for (i,j),val in zip(X,y))
        r2_partial_resid = sum([(p - val)**2 for p,val in zip(partial_preds, y)])
        r2_resid = sum([(p - val)**2 for p,val in zip(preds, y)])
        print(1 - (float(r2_partial_resid) / float(r2)))
        print(1 - (float(r2_resid) / float(r2)))

    def to_csv(self, row_idx2name = None, col_idx2name = None, stdout=False):
        # row_features, col_features = pred.row_features, pred.col_features #train(N1, N2, vals, feature_cnt, iter_cnt=iterations)
        if not row_idx2name:
            row_idx2name = dict((i,i) for i in range(self.N1))
        if not col_idx2name:
            col_idx2name = dict((i,i) for i in range(self.N2))
        SORT_BY = "feature_0"

        user_cols = ["user","offset"] + ["feature_"+str(k) for k,_ in enumerate(self.row_features)]
        df_users = pd.DataFrame([dict([("user",row_idx2name[i]), ("offset",self.row_offsets[i])] + [("feature_"+str(k),f) for k,f in enumerate(features)]) for i,features in enumerate(zip(*self.row_features))],columns=user_cols).sort_values(by=SORT_BY)

        item_cols = ["item","offset"] + ["feature_"+str(k) for k,_ in enumerate(self.col_features)]
        df_items = pd.DataFrame([dict([("items",col_idx2name[i]), ("offset",self.col_offsets[i])] + [("feature_"+str(k),f) for k,f in enumerate(features)]) for i,features in enumerate(zip(*self.col_features))], columns=item_cols).sort_values(by=SORT_BY)

        #replace small values with 0
        for i,c in enumerate([c for c in df_users.columns if not c == "user"]):
            try:
                df_users.loc[abs(df_users[c]) < 1e-3,c] = 0
            except Exception as e:
                raise e
        for i,c in enumerate([c for c in df_items.columns if not c == "item"]):
            df_items.loc[abs(df_items[c]) < 1e-3,c] = 0

        # df_users.to_csv(sys.stdout, index=False)
        # print "---------------------\n\n\n"
        # df_items.to_csv(sys.stdout, index=False)
        if stdout:
            df_users.to_csv(sys.stdout, index=False)
            df_items.to_csv(sys.stdout, index=False)
        else:
            df_users.to_csv("/tmp/pca_users.csv", index=False)
            df_items.to_csv("/tmp/pca_items.csv", index=False)



def train(N1, N2, vals, feature_cnt, iter_cnt=10000):
    row_features = [[np.random.normal(scale=0.1, size=1)[0] for i in range(N1)] for k in range(feature_cnt)]
    col_features = [[np.random.normal(scale=0.1, size=1)[0] for i in range(N2)] for k in range(feature_cnt)]
    lrate = 0.001
    for it in range(iter_cnt):
        print(f"Iteration {it}")
        if it%100 == 0:
            total_err = 0
            for i,j,v in vals:
                total_err += abs(v - sum([row_features[k][i]*col_features[k][j] for k in range(feature_cnt)]))
            print("total_err: " + str(total_err))
        for f in range(feature_cnt):
            for i,j,v in vals:
                err = v - sum([row_features[k][i]*col_features[k][j] for k in range(feature_cnt)])
                row_val = row_features[f][i]
                col_val = col_features[f][j]
                row_features[f][i] += lrate * err * col_val
                col_features[f][j] += lrate * err * row_val
    return row_features, col_features


def test(i,j,v):
    N1 = 3
    N2 = 3
    mean = 0
    row_means = [0] * N1
    col_means = [0] * N2
    feature_cnt = 1
    err = (v - (mean + row_means[i] + col_means[j] + sum([k1*k2 for k1,k2 in zip(row_factors[i],col_factors[j])])))
    #scipy.optimize.minimize(fn1, data1, jac=grad_fn1, args=tuple(args_dict.values()), method=method, bounds=bounds1, options=options1)

def readCL():
    parser = argparse.ArgumentParser()
    parser.add_argument("-f","--infile")
    parser.add_argument("-i","--iterations",type=int, default=5000)
    args = parser.parse_args()
    return args.infile, args.iterations

def test_filter_pca():
    N1 = 5
    N2 = 5
    feature_cnt = 2
    vals = [(i,j,i*j + np.random.normal(scale=0.1, size=1)[0]) for i in range(N1) for j in range(N2)]
    X = [v[:2] for v in vals]
    y = [v[2] for v in vals]
    pred = FilterPca(feature_cnt=feature_cnt, N1=N1, N2=N2)
    pred.fit(X,y,1000)
    pred.to_csv(stdout=True)
    # row_features, col_features = tf_train(N1, N2, vals, feature_cnt, 10000)
    # print row_features, col_features



def tf_train(N1, N2, vals, feature_cnt, iter_cnt):
    #some pieces from here:
    #http://katbailey.github.io/post/matrix-factorization-with-tensorflow/

    item_row = tf.placeholder(tf.int32, [None])
    item_col = tf.placeholder(tf.int32, [None])
    y_ = tf.placeholder(tf.float32, [None])

    total_mean = tf.Variable(tf.random_normal([],stddev=0.1))
    row_offsets = tf.Variable(tf.random_normal([N1], stddev=0.1))
    col_offsets = tf.Variable(tf.random_normal([N2], stddev=0.1))
    row_features = tf.Variable(tf.random_normal([N1, feature_cnt], stddev=0.1))
    col_features = tf.Variable(tf.random_normal([N2, feature_cnt], stddev=0.1))
    regularization_penalty = tf.constant(0e-4)

    item_row_offset = tf.gather(row_offsets, item_row)
    item_col_offset = tf.gather(col_offsets, item_col)

    item_row_features = tf.gather(row_features, item_row)
    item_col_features = tf.gather(col_features, item_col)
    # item_row_features = tf.slice(row_features, [item_row,0], [1, -1])
    # item_col_features = tf.slice(col_features, [item_col,0], [1, -1])
    # item_row_features = tf.gather(tf.reshape(result, [-1]), user_indices * tf.shape(result)[1] + item_indices, name="extract_training_ratings")

    # row_features = tf.Print(row_features,[row_features], "row_features: ", summarize=1000)
    # item_row_features = tf.Print(item_row_features,[item_row_features], "item_row_features: ")

    feature_prod = tf.diag_part(tf.matmul(item_row_features, item_col_features, transpose_b = True))
    pred = tf.add_n([tf.fill(tf.shape(item_row), total_mean), item_row_offset, item_col_offset, feature_prod])
    error = tf.reduce_mean(tf.square(tf.sub(pred, y_)))
    # reg_penalty = tf.mul(regularization_penalty,
    #                          tf.nn.l2_loss(tf.concat(0,[tf.expand_dims(total_mean, 0),
    #                                                     tf.reshape(row_offsets, [-1]),
    #                                                     tf.reshape(col_offsets, [-1]),
    #                                                     tf.reshape(row_features, [-1]),
    #                                                     tf.reshape(col_features, [-1])])))

    reg_penalty = tf.mul(regularization_penalty,
                             tf.reduce_sum(tf.abs((tf.concat(0,[tf.expand_dims(total_mean, 0),
                                                        tf.reshape(row_offsets, [-1]),
                                                        tf.reshape(col_offsets, [-1]),
                                                        tf.reshape(row_features, [-1]),
                                                        tf.reshape(col_features, [-1])])))))


    cost = tf.add(error, reg_penalty)
    cost_summary = tf.scalar_summary("cost", cost)

    tf.scalar_summary("reg_penalty",reg_penalty)
    tf.scalar_summary("error",error)
    tf.scalar_summary("cost",cost)

    train_step = tf.train.AdamOptimizer(1e-3).minimize(cost)

    sess = tf.Session()
    sess.run(tf.initialize_all_variables())

    summary_op = tf.merge_all_summaries()
    writer = tf.train.SummaryWriter("/tmp/recommender_logs", sess.graph)

    print(len(vals))
    for i in range(iter_cnt):
        training_set, test_set = gen_train_test(vals, 0.8)
        item_row_val, item_col_val, y_val = training_set[i % len(training_set)]
        _, cost_val, error_val = sess.run([train_step, cost, error], {item_row: [item_row_val],
                                               item_col: [item_col_val],
                                               y_: [y_val]})
        if i % 500 == 0:
            print("training cost/error: " + str(cost_val) + "/" + str(error_val))
            rows, cols, labels = zip(*test_set)
            out = sess.run([total_mean, row_offsets, col_offsets, row_features, col_features, pred, reg_penalty, cost, error],
                           {item_row: rows,
                            item_col: cols,
                            y_: labels})
            # print("info: " + str(out))
            # print("pred: " + str(out[5]))
            # print("reg: " + str(out[6]))
            # print("cost: " + str(out[7]))
            print("test error: " + str(out[8]))
            # print("info: " + str([out[0],out[5]]))
    out = sess.run([row_features, col_features, row_offsets, col_offsets])
    row_features_out, col_features_out, row_offsets_out, col_offsets_out = out
    return [row_offsets_out] + zip(*row_features_out), [col_offsets_out] + zip(*col_features_out)

def gen_train_test(l, training_prob):
    train = []
    test = []
    for i in l:
        if random.random() < training_prob:
            train.append(i)
        else:
            test.append(i)
    return train, test


def mean(l):
    return sum(l) / float(len(l))

def preprocess():
    infile, iterations = readCL()
    if not infile: raise
    print("loading...")
    df = pd.read_csv(infile)
    all_users = list(set(df.iloc[:,0].values))
    all_items = list(set(df.iloc[:,1].values))
    # items_to_keep = ["q156917", "q210", "q118236", "q42"]
    # items_to_keep = set(random.sample(all_items,1000))
    # users_to_keep = set(random.sample(all_users,1000))
    items_to_keep = set(all_items)
    users_to_keep = set(all_users)

    print("filtering")
    df = df[df.apply(lambda x: x[0] in users_to_keep and x[1] in items_to_keep, axis=1)]
    print("done filtering")
    # df.to_csv("/tmp/a.csv", index=False)
    # raise
    users = list(set(df.iloc[:,0].values))
    items = list(set(df.iloc[:,1].values))
    user2idx = dict((a,i) for i,a in enumerate(users))
    item2idx = dict((q,i) for i,q in enumerate(items))
    idx2user = dict((i,a) for i,a in enumerate(users))
    idx2item = dict((i,q) for i,q in enumerate(items))
    N1 = len(users)
    N2 = len(items)
    vals = [[user2idx[a], item2idx[q], v] for a,q,v in df.values]
    vals_train, vals_test = gen_train_test(vals, training_prob = 0.8)
    X_train = [v[:2] for v in vals_train]
    y_train = [v[2] for v in vals_train]
    X_test = [v[:2] for v in vals_test]
    y_test = [v[2] for v in vals_test]
    return idx2user, idx2item, N1, N2, X_train, y_train, X_test, y_test

def main():
    # total_mean = mean([v for _,_,v in vals])
    # vals = [(i,j,v-total_mean) for i,j,v in vals]
    # item_means = dict((j,mean([v for _,q,v in vals if q == j])) for j in range(N2))
    # vals = [(i,j,v-item_means[j]) for i,j,v in vals]
    idx2user, idx2item, N1, N2, X_train, y_train, X_test, y_test = preprocess()
    feature_cnt = 2
    MIN_VAL = -1
    MAX_VAL = 1
    pred = FilterPca(feature_cnt=feature_cnt, N1=N1, N2=N2, minval=MIN_VAL, maxval=MAX_VAL)
    print("prefit")
    for i in range(10):
        pred.fit(X_train, y_train, 100)
        pred.score(X_train, y_train)
        pred.score(X_test, y_test)
        print(sum(abs(x) for x in pred.col_features[0]))
        print(sum(abs(x) for x in pred.row_features[0]))
        pred.to_csv(idx2user, idx2item)

if __name__ == "__main__":
    main()
    # test_filter_pca()

File: filter_pca.cpp
#include <stdio.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include <tuple>
#include <vector>
#include <cmath>
#include <random>
#include <set>
#include <map>

typedef std::tuple<
  std::map<int,std::string>,
  std::map<int,std::string>,
  int,
  int,
  std::vector<std::vector<int>>,
  std::vector<std::vector<int>>> preprocess_tuple;
typedef std::vector<std::vector<std::string>> dataframe;

template<typename T>
std::tuple<std::vector<T>, std::vector<T>> gen_train_test(std::vector<T> data, float training_prob) {
  std::vector<T> train;
  std::vector<T> test;
  for (T row : data) {
    float r = ((float) rand() / (RAND_MAX));
    if (r < training_prob) {
      train.push_back(row);
    } else {
      test.push_back(row);
    }
  }
  return std::tuple<std::vector<T>,
                    std::vector<T>>(train, test);
}

preprocess_tuple preprocess(std::vector<std::vector<std::string>> &data) {
  std::set<std::string> users;
  std::set<std::string> questions;
  for (std::vector<std::string> row: data) {
    std::string user = row[0];
    std::string question = row[1];
    std::string val   = row[2];
    users.insert(user);
    questions.insert(question);
  }
  std::vector<std::string> user_vec(users.begin(), users.end());
  std::vector<std::string> question_vec(questions.begin(), questions.end());

  std::map<int,std::string> idx2user;
  std::map<int,std::string> idx2question;
  std::map<std::string,int> user2idx;
  std::map<std::string,int> question2idx;
  for (int i=0; i<user_vec.size(); i++) {
    std::string u = user_vec[i];
    idx2user[i] = u;
    user2idx[u] = i;
  }
  for (int i=0; i<question_vec.size(); i++) {
    std::string q = question_vec[i];
    idx2question[i] = q;
    question2idx[q] = i;
  }

  std::vector<std::vector<std::string>> filtered_data;
  for (std::vector<std::string> row: data) {
    std::string user = row[0];
    std::string question = row[1];
    std::string val   = row[2];
    int user_id = user2idx[user];
    if (user_id < 1000) {
      filtered_data.push_back(row);
    }
  }

  std::tuple<std::vector<std::vector<std::string>>,
             std::vector<std::vector<std::string>>> out = gen_train_test<std::vector<std::string>>(filtered_data, 0.8);
  std::vector<std::vector<std::string>> train = std::get<0>(out);
  std::vector<std::vector<std::string>> test = std::get<1>(out);
  std::vector<std::vector<int>> train_int;
  std::vector<std::vector<int>> test_int;

  for (std::vector<std::string> row: train) {
    std::vector<int> out;
    out.push_back(user2idx[row[0]]);
    out.push_back(question2idx[row[1]]);
    out.push_back(std::stoi(row[2]));
    train_int.push_back(out);
  }
  for (std::vector<std::string> row: test) {
    std::vector<int> out;
    out.push_back(user2idx[row[0]]);
    out.push_back(question2idx[row[1]]);
    out.push_back(std::stoi(row[2]));
    test_int.push_back(out);
  }

  return preprocess_tuple(idx2user, idx2question,
                          user_vec.size(), question_vec.size(),
                          train_int, test_int);
}

void printData(std::vector<std::vector<int>> &data) {
  for (std::vector<int> row: data) {
    for (int field: row) {
      printf("%i\n",field);
    }
  }
}

void print_vector(std::vector<std::string> &v, std::ostream & os) {
  std::string out = "";
  for (std::string s : v) {
    out += s + ",";
  }
  out = out.substr(0,out.size() - 1); //drop the trailing comma
  os << out.c_str() << std::endl;
}

void print_2d_vector(std::vector<std::vector<std::string>> &v, std::ostream & os = std::cout) {
  for (std::vector<std::string> v2: v) {
    print_vector(v2, os);
  }
}

std::vector<std::vector<std::string>> loadCsv(std::string filename) {
  std::ifstream infile(filename);
  std::string line;
  std::vector<std::vector<std::string>> data;
  while (std::getline(infile, line)) {
    std::stringstream lineStream(line);
    std::string field;
    std::vector<std::string> row;
    while (std::getline(lineStream,field,',')) {
      row.push_back(field);
    }
    if (row.size() == 3 && row[2] != "val") {
      data.push_back(row);
    }
  }
  // printData(data);
  return data;
}

float pred(int rowId, int colId, float total_mean, std::vector<float> &col_means, std::vector<std::vector<float>> &row_features, std::vector<std::vector<float>> &col_features) {
  float out = total_mean + col_means[colId];
  for (int i=0; i<row_features[0].size(); i++) {
    out += row_features[rowId][i] * col_features[colId][i];
  }
  return out;
}

void ratingPca(std::string filename) {
  int dim = 5;
  int matrix[dim][dim];
  for (int i=0; i<dim; i++) {
    for (int j=0; j<dim; j++) {
      matrix[i][j] = i * j;
    }
  }

  // printf("%d\n",matrix[2][2]);

  printf("Loading csv...\n");
  std::vector<std::vector<std::string>> raw_data = loadCsv(filename);

  printf("Preprocessing data...\n");
  preprocess_tuple data = preprocess(raw_data);

  std::map<int, std::string> idx2user = std::get<0>(data);
  std::map<int, std::string> idx2question = std::get<1>(data);
  int row_cnt = std::get<2>(data);
  int col_cnt = std::get<3>(data);
  std::vector<std::vector<int>> train = std::get<4>(data);
  std::vector<std::vector<int>> test = std::get<5>(data);

  int feature_cnt = 1;

  std::vector<float> col_sums(col_cnt,0);
  std::vector<float> col_cnts(col_cnt,0);

  int total_cnt = 0;
  float total_sum = 0;

  printf("Computing column means...\n");
  for(std::vector<int> row: train) {
    int rowId = row[0];
    int colId = row[1];
    int val   = row[2];
    col_sums[colId] += val;
    col_cnts[colId]++;
    total_sum += val;
    total_cnt++;
  }

  float total_mean = total_sum / total_cnt;

  //setup col_means
  std::vector<float> col_means(col_cnt,0);
  for(int i = 0; i<col_sums.size(); i++) {
    if (col_cnts[i] > 0) {
      col_means[i] = (col_sums[i] / col_cnts[i]) - total_mean;
    } else {
      col_means[i] = 0.0;
    }
    // printf("%i,%f\n",i,col_means[i]);
  }

  //initialize features to 0 + random noise
  std::random_device rd;
  std::mt19937 gen(rd());
  std::normal_distribution<float> d(0,0.01);

  std::vector<std::vector<float>> row_features;
  std::vector<std::vector<float>> col_features;

  printf("Initializing features...\n");
  for (int r=0; r<row_cnt; r++) {
    std::vector<float> row;
    for (int k=0; k<feature_cnt; k++) {
      row.push_back(d(gen));
    }
    row_features.push_back(row);
  }
  for (int c=0; c<col_cnt; c++) {
    std::vector<float> col;
    for (int k=0; k<feature_cnt; k++) {
      col.push_back(d(gen));
    }
    col_features.push_back(col);
  }

  // row_cnt,std::vector<float>(feature_cnt,0));
  // col_cnt,std::vector<float>(feature_cnt,0));

  int epoch_cnt = 10000;
  float lrate = 0.001; //0.01
  float lam = 15; //.1; //7;

  printf("Training...\n");
  for (int i=0; i<epoch_cnt; i++) {
    if (i % 100 == 0) {
      //compute training set error
      float sum_sq_err = 0;
      for (std::vector<int> row: train) {
        int rowId = row[0];
        int colId = row[1];
        int val   = row[2];
        float err = val - pred(rowId, colId, total_mean, col_means, row_features, col_features);
        sum_sq_err += pow(err,2);
      }
      float obj_fn = sum_sq_err;
      for (int k=0; k<feature_cnt; k++) {
        for (int c=0; c<col_features.size(); c++) {
          obj_fn += lam * pow(pow(col_features[c][k],2),1);
        }
        for (int r=0; r<row_features.size(); r++) {
          obj_fn += lam * pow(pow(row_features[r][k],2),1);
        }
      }
      printf("-----\n");
      printf("sum_sq_err: %f\n", sum_sq_err);
      printf("obj_fn: %f\n", obj_fn);

      //compute test set error
      float test_sum_sq_err = 0;
      for (std::vector<int> row: test) {
        int rowId = row[0];
        int colId = row[1];
        int val = row[2];
        float err = val - pred(rowId, colId, total_mean, col_means, row_features, col_features);
        test_sum_sq_err += pow(err, 2);
      }
      printf("test sum_sq_err: %f\n", test_sum_sq_err);
    }
    for (int k=0; k<feature_cnt; k++) {
      for (std::vector<int> row: train) {
        int rowId = row[0];
        int colId = row[1];
        int val   = row[2];
        float err = val - pred(rowId, colId, total_mean, col_means, row_features, col_features);
        float row_val = row_features[rowId][k];
        float col_val = col_features[colId][k];
        // if (i > 50000 && rowId == 0) {
        //   printf("%i,%i,%i,%f\n", rowId, colId, val, err);
        //   printf("%f,%f\n", row_val, col_val);
        //   printf("%f,%f\n", lrate * (err * col_val - lam * row_val), lrate * (err * row_val - lam * col_val));
        // }
        row_features[rowId][k] += lrate * (err * col_val);
        col_features[colId][k] += lrate * (err * row_val);
      }
      for (int rowId=0; rowId < row_features.size(); rowId++) {
        float row_val = row_features[rowId][k];
        row_features[rowId][k] -= lrate * lam * row_val;
        // if (row_val != 0) {
        //   row_features[rowId][k] -= (row_val > 0 ? 1 : -1) * lam;
        // }
      }
      for (int colId=0; colId < col_features.size(); colId++) {
        float col_val = col_features[colId][k];
        col_features[colId][k] -= lrate * lam * col_val;
        // if (col_val != 0) {
        //   col_features[colId][k] -= (col_val > 0 ? 1 : -1) * lam;
        // }
      }
    }
  }

  // printf("col means:\n");
  // print_vector(col_means);
  // printf("row features:\n");
  // print_2d_vector(row_features);
  // printf("col features:\n");
  // print_2d_vector(col_features);


  printf("Writing to csv...\n");

  //write to csv
  std::vector<std::string> user_cols;
  user_cols.push_back("user");
  for (int k=0; k<feature_cnt; k++) {
    user_cols.push_back("feature_" + std::to_string(k));
  }
  dataframe df_users;
  df_users.push_back(user_cols);
  for (int i=0; i<row_features.size(); i++) {
    std::vector<std::string> row;
    row.push_back(idx2user[i]);
    for (int k=0; k<feature_cnt; k++) {
      row.push_back(std::to_string(row_features[i][k]));
    }
    df_users.push_back(row);
  }

  std::vector<std::string> question_cols;
  question_cols.push_back("question");
  for (int k=0; k<feature_cnt; k++) {
    question_cols.push_back("feature_" + std::to_string(k));
  }
  dataframe df_questions;
  df_questions.push_back(question_cols);
  for (int i=0; i<col_features.size(); i++) {
    std::vector<std::string> row;
    row.push_back(idx2question[i]);
    for (int k=0; k<feature_cnt; k++) {
      row.push_back(std::to_string(col_features[i][k]));
    }
    df_questions.push_back(row);
  }

  std::ofstream of1("/tmp/b.csv");
  print_2d_vector(df_users, of1);

  std::ofstream of2("/tmp/c.csv");
  print_2d_vector(df_questions, of2);
}

int main(int argc, const char** argv) {
  printf("helloworld\n");

  std::string csvFile = "/home/jtrigg/files/okcupid/sample_data_test.csv";

  if (argc > 1) {
    csvFile = std::string(argv[1]);
  }

  ratingPca(csvFile);
}
